repo/blob/azure/azure_storage.go

```diff
@@ -226,6 +293,203 @@ func (az *azStorage) DisplayName() string {
+	_, err := az.service.ServiceClient().
+		NewContainerClient(az.container).
+		NewBlobClient(az.getObjectNameString(originalBlob)).
+		GetProperties(ctx, nil)
+
+	return errors.Is(translateError(err), blob.ErrBlobNotFound)
+}
+
+func (az *azStorage) getBlobName(it *azblobmodels.BlobItem) string {
+	n := *it.Name
+	return n[len(az.Prefix):]
+}
+
+func (az *azStorage) getBlobMeta(it *azblobmodels.BlobItem) blob.Metadata {
+	bm := blob.Metadata{
+		BlobID: blob.ID(az.getBlobName(it)),
+		Length: *it.Properties.ContentLength,
+	}
+
+	// see if we have 'Kopiamtime' metadata, if so - trust it.
+	if t, ok := timestampmeta.FromValue(stringDefault(it.Metadata["kopiamtime"], "")); ok {
+		bm.Timestamp = t
+	} else {
+		bm.Timestamp = *it.Properties.LastModified
+	}
+	return bm
+}
+
+// cleanupParallel removes files that have a matching delete marker file, providing the retention period has passed.
+// If the original file has been deleted then the delete marker file is also deleted, providing the retention period has passed.
+func (az *azStorage) cleanupParallel(ctx context.Context, logger logging.Logger, deleteMarkerFiles []blob.Metadata) (deletedBlobs, deletedMarkers int, err error) {
+	var (
+		parallel          = runtime.NumCPU()
+		wg                sync.WaitGroup
+		deletedBlobsCtr   = new(uint32)
+		deletedMarkersCtr = new(uint32)
+		errChan           = make(chan error, parallel)
+	)
+
+	deleteMarkerBlobs := make(chan blob.Metadata, len(deleteMarkerFiles))
+
+	for i := 0; i < parallel; i++ {
+		wg.Add(1)
+
+		go func() {
+			defer wg.Done()
+
+			for bm := range deleteMarkerBlobs {
+				originalFile, ok := getOriginalFile(string(bm.BlobID))
+				if !ok {
+					continue
+				}
+				blobsDeleted, fileRemoved, err := az.attemptFileDeletion(ctx, blob.ID(originalFile))
+				if err != nil {
+					errChan <- errors.Wrap(err, "failed to delete original blob")
+					return
+				}
+				atomic.AddUint32(deletedBlobsCtr, uint32(blobsDeleted))
+				if !fileRemoved {
+					logger.Debugf("skipped cleanup of file %s, too early to delete", originalFile)
+					continue
+				}
+
+				blobsDeleted, fileRemoved, err = az.attemptFileDeletion(ctx, bm.BlobID)
+				if err != nil {
+					errChan <- errors.Wrap(err, "failed to delete original blob")
+					return
+				}
+				if !fileRemoved {
+					logger.Debugf("skipped cleanup of file %s, too early to delete", string(bm.BlobID))
+				}
+				atomic.AddUint32(deletedMarkersCtr, uint32(blobsDeleted))
+			}
+		}()
+	}
+
+	for _, bm := range deleteMarkerFiles {
+		deleteMarkerBlobs <- bm
+	}
+
+	close(deleteMarkerBlobs)
+
+	wg.Wait()
+	close(errChan)
+
+	return int(*deletedBlobsCtr), int(*deletedMarkersCtr), <-errChan
+}
+
+// attemptFileDeletion tries to delete a blob if it still exists and its retention period has passed.
+func (az *azStorage) attemptFileDeletion(ctx context.Context, b blob.ID) (deletedBlobs int, fileRemoved bool, err error) {
+	_, err = az.service.DeleteBlob(ctx, az.container, az.getObjectNameString(b), nil)
```

Comment:
Just to make sure: this permanently deletes all previous versions of the blob in one operation, correct?